src/Adapters/Etw.Net451/EtwTelemetryModule.cs

```diff
@@ -0,0 +1,247 @@
+                return;
+            }
+
+            if (this.isDisposed)
+            {
+                errorMessage = "Can't initialize a module that is disposed. The initialization is terminated.";
+                EventSourceListenerEventSource.Log.ModuleInitializationFailed(
+                    nameof(EtwTelemetryModule),
+                    errorMessage);
+                return;
+            }
+
+            bool? isProcessElevated = this.traceEventSession.IsElevated();
+            if (!isProcessElevated.HasValue || !isProcessElevated.Value)
+            {
+                errorMessage = "The process is required to be elevated to enable ETW providers. The initialization is terminated.";
+                EventSourceListenerEventSource.Log.ModuleInitializationFailed(
+                    nameof(EtwTelemetryModule),
+                    errorMessage);
+                
+                // Throws so that user will be able to see the exception message in Output Window for debugging.
+                throw new UnauthorizedAccessException(errorMessage);
+            }
+
+            lock (this.lockObject)
+            {
+                this.client = new TelemetryClient(configuration);
+
+                // sdkVersionIdentifier will be used in telemtry entry as a identifier for the sender.
+                // The value will look like: etw:x.x.x-x
+                const string SdkVersionIdentifier = "etw:";
+                this.client.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion(SdkVersionIdentifier);
+
+                if (this.isInitialized)
+                {
+                    this.isInitialized = false;
+                    this.DisableProviders();
+                    this.enabledProviderIds.Clear();
+                    this.enabledProviderNames.Clear();
+                }
+
+                if (this.Sources.Count == 0)
+                {
+                    EventSourceListenerEventSource.Log.NoSourcesConfigured(moduleName: nameof(EtwTelemetryModule));
+                    return;
+                }
+
+                this.EnableProviders();
+                try
+                {
+                    // Start the trace session
+                    Task.Factory.StartNew(() => this.StartTraceEventSession(this.traceEventSession, this.client), TaskCreationOptions.LongRunning);
+                }
+                finally
+                {
+                    this.isInitialized = true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposes the module.
+        /// </summary>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes the module.
+        /// </summary>
+        /// <param name="isDisposing">Indicate if it is called by Dispose().</param>
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (this.isDisposed)
```

Comment:
My point is you need to decide whether it's safe to call `Dispose` of a child twice or not. If it's sage - no need for this logic. If not - we have to avoid "ghost bug chasing" by introducing locks